### PR TITLE
WebDav:listFolder - skip adding phantom items

### LINKED ISSUE #12025: Cloud Storage repetitive phantom folder
* KOReader version: 2024.04-161
* Device: Kindle Scribe (5.16.2)

#### Issue
When I go to Yandex.Disk WebDav cloud storage, then enter folder `Books`, there is a duplicate phantom folder in that folder with the same name `Books`. See below.

FileManager_2024-06-13_164051

#### Steps to reproduce
In the crash log below:
- Open KOReader
- Exit from Book
- Go to Cloud Storage
- Click `Books`
- Exit from cloud
- Exit KOReader.

##### `crash.log` (if applicable)
[cloud_repetitive_phantom_folder.log](https://github.com/user-attachments/files/15828597/cloud_repetitive_phantom_folder.log)

### PR DESCRIPTION
**Pull request by mergen3107**
Closes #12025 

This is my try to fix the issue when a WebDav folder contains a phantom item with exact same name as that of the current folder itself. This phantom item is, of course, non-interactable.

So, in the `for` loop where all url-decoded items are added to the `webdav_list` table, I check whether current-iteration item is phantom ("path to current item with trailing slash" is equal to "path of current folder") or real.

Any suggestions or comments are appreciated! :D

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12034)

### CODE REVIEW

**Comment by mergen3107** [comment]
I tried to avoid solutions with `table.remove` for two reasons: I don't want to blindly remove just first items (what if they are real?) and also this operation duplicates a whole table lest the first item (what if that table is huge?).

**Comment by mergen3107** [comment]
Tested on emulator 2024.04-186 with my cloud settings as on Kindle Scribe.

**Comment by mergen3107** [comment]
Found two issues:
1) Phantom folder appears if I go to a folder with spaces in the name
2) Phantom folder appears if I go to next level of the folder. E.g. `Books/` doesn't have the phantom item, but `Books/123/` does.

**Comment by mergen3107** [comment]
First case above happens because `path` (of current folder) has `%20` in lieu of a white space. So `/Deus Ex PC Build` string is `/Deus%20Ex%20PC%20Build`. Is this a results of html decode?

**Comment by mergen3107** [comment]
Second problems happens because I compare two different things now:
```
06/15/24-23:13:25 DEBUG LOGG item_name =  DeDRM 
06/15/24-23:13:25 DEBUG LOGG path =  /Kindle/DeDRM
```
I should probably strip all characters until including last slash.

**Comment by mergen3107** [comment]
Number 1 is fixed by using unencoded `folder_path` instead of `path`.
Number 2 is fixed by comparing `item_name` with a string shortened from full `folder_path` by the length of `item_name`.
For example, item_name = `123 456`, folder_path = `/Books/123 456`, and shorted string to compare `item_name` with is `123 456`.

**Comment by mergen3107** [comment]
I know it looks way too hacky, any suggestions are welcome!

Otherwise, emulator tests are great!

**Comment by NiLuJe** [comment]
It seems like `isCurrentDirectory` is what should be taking care of this.

It also seems that this is seven billion times simpler, which I like.

Then again, I'm not familiar with DAV, so I'm unsure of whatever edge-cases it's currently trying to deal with.

TL;DR: Does things horribly break with a simple

```diff
diff --git a/frontend/apps/cloudstorage/webdavapi.lua b/frontend/apps/cloudstorage/webdavapi.lua
index f0d36696f..ccb584684 100644
--- a/frontend/apps/cloudstorage/webdavapi.lua
+++ b/frontend/apps/cloudstorage/webdavapi.lua
@@ -125,11 +125,12 @@ function WebDavApi:listFolder(address, user, pass, folder_path, folder_mode)
             if string.sub( item_fullpath, -1 ) == "/" then
                 item_fullpath = string.sub( item_fullpath, 1, -2 )
             end
-            local is_current_dir = self:isCurrentDirectory( util.urlDecode(item_fullpath), address, folder_path )
 
             local item_name = util.urlDecode( FFIUtil.basename( item_fullpath ) )
             item_name = util.htmlEntitiesToUtf8(item_name)
 
+            local is_current_dir = item_name == string.sub(folder_path, -#item_name)
+
             local is_not_collection = item:find("<[^:]*:resourcetype/>") or
                 item:find("<[^:]*:resourcetype></[^:]*:resourcetype>")
```

**Comment by mergen3107** [comment]
@NiLuJe 
Things didn't break, but they didn't work neither - phantom folders are back!

**Comment by mergen3107** [comment]
Ah, did you mean, this diff is applied **on top** of this PR? I applied it on fresh nightly

**Comment by mergen3107** [comment]
> applied **on top** of this PR

@NiLuJe 
All good in this case (on Kindle). I'll make a commit, thanks!

**Comment by mergen3107** [comment]
Sorry, I think I got it finally (need some sleep).
Let me push commit with comments

**Comment by mergen3107** [comment]
@NiLuJe 
With last commit, it works the same good as original PR :D nothing broke

**Comment by mergen3107** [comment]
Overall diff looks much cleaner, thank you!

**Comment by mergen3107** [comment]
Tested on emulator as well, all good.

**Comment by mergen3107** [comment]
(removing those lines made the phantom folders come back)
```
-            if string.sub( item_fullpath, -1 ) == "/" then
-                item_fullpath = string.sub( item_fullpath, 1, -2 )
-            end
```
It works in the current state of this PR.

**Comment by hius07** [comment]
Dunno anything about the issue, but after your changes `isCurrentDirectory()` is not used anymore.
It is worth to investigate why rather complicated method was needed, instead of just one string comparison.
Maybe, some edge cases or something?

**Comment by NiLuJe** [comment]
> Dunno anything about the issue, but after your changes `isCurrentDirectory()` is not used anymore.

That's possibly going to annoy Luacheck, for the record. I didn't mention that before ;p.

> It is worth to investigate why rather complicated method was needed, instead of just one string comparison.
Maybe, some edge cases or something?

Yep, I'm not familiar with the codepaths involved (or DAV) to say, but it is a concern.

**Comment by NiLuJe** [comment]
> (removing those lines made the phantom folders come back)
> 
> ```
> -            if string.sub( item_fullpath, -1 ) == "/" then
> -                item_fullpath = string.sub( item_fullpath, 1, -2 )
> -            end
> ```
> 
> It works in the current state of this PR.

Huh.

What if we stop doing stupid things like passing URL-encoded crap to `basename`?

i.e., `local item_name = FFIUtil.basename(util.urlDecode(item_fullpath))`

**Comment by NiLuJe** [comment]
No context braindump because the torturous logic made my head hurt:

```diff
diff --git a/frontend/apps/cloudstorage/webdavapi.lua b/frontend/apps/cloudstorage/webdavapi.lua
index f0d36696f..b76f487bd 100644
--- a/frontend/apps/cloudstorage/webdavapi.lua
+++ b/frontend/apps/cloudstorage/webdavapi.lua
@@ -12,44 +12,23 @@ local lfs = require("libs/libkoreader-lfs")
 local WebDavApi = {
 }
 
-function WebDavApi:getJoinedPath( address, path )
-    local path_encoded = self:urlEncode( path ) or ""
-    local address_strip = address:sub(-1) == "/" and address:sub(1, -2) or address
-    local path_strip = path_encoded:sub(1, 1) == "/" and path_encoded:sub(2) or path_encoded
-    return address_strip .. "/" .. path_strip
+-- Trim leading & trailing character `c` from string `s` (based on util.trim)
+function WebDavApi.trim(s, c)
+    local from = s:match"^"..c.."*()"
+    return from > #s and "" or s:match(".*".."[^"..c.."]", from)
 end
 
-function WebDavApi:isCurrentDirectory( current_item, address, path )
-    local is_home, is_parent
-    local home_path
-    -- find first occurence of / after http(s)://
-    local start = string.find( address, "/", 9 )
-    if not start then
-        home_path = "/"
-    else
-        home_path = string.sub( address, start )
+-- Trim trailing character `c` from string `s` (based on util.rtrim)
+function WebDavApi.rtrim(s, c)
+    local n = #s
+    while n > 0 and s:find("^"..c, n) do
+        n = n - 1
     end
-    local item
-    if string.sub( current_item, -1 ) == "/" then
-        item = string.sub( current_item, 1, -2 )
-    else
-        item = current_item
-    end
-
-    if item == home_path then
-        is_home = true
-    else
-        local temp_path = string.sub( item, string.len(home_path) + 1 )
-        if string.sub( path, -1 ) == "/" then path = string.sub( path, 1, -2 ) end
-        if temp_path == path then
-            is_parent = true
-        end
-    end
-    return is_home or is_parent
+    return s:sub(1, n)
 end
 
--- version of urlEncode that doesn't encode the /
-function WebDavApi:urlEncode(url_data)
+-- Variant of util.urlEncode that doesn't encode the /
+function WebDavApi.urlEncode(url_data)
     local char_to_hex = function(c)
         return string.format("%%%02X", string.byte(c))
     end
@@ -60,28 +39,29 @@ function WebDavApi:urlEncode(url_data)
     return url_data
 end
 
+function WebDavApi:getJoinedPath(address, path)
+    local path_encoded = self.urlEncode(path) or ""
+    -- Strip leading & trailing slashes from `path`
+    local sane_path = self.trim(path_encoded, "/")
+    -- Strip trailing slashes from address for now
+    local sane_address = self.rtrim(address, "/")
+    -- Join our final URL
+    return sane_address .. "/" .. sane_path
+end
+
 function WebDavApi:listFolder(address, user, pass, folder_path, folder_mode)
-    local path = self:urlEncode( folder_path )
+    local path = self.urlEncode(folder_path) or ""
     local webdav_list = {}
     local webdav_file = {}
 
-    local has_trailing_slash = false
-    local has_leading_slash = false
-    if string.sub( address, -1 ) == "/" then has_trailing_slash = true end
-    if path == nil or path == "/" then
-        path = ""
-    elseif string.sub( path, 1, 1 ) == "/" then
-        if has_trailing_slash then
-            -- too many slashes, remove one
-            path = string.sub( path, 2 )
-        end
-        has_leading_slash = true
-    end
-    if not has_trailing_slash and not has_leading_slash then
-        address = address .. "/"
-    end
-    local webdav_url = address .. path
-    if string.sub(webdav_url, -1) ~= "/" then
+    -- Strip leading & trailing slashes from `path`
+    path = self.trim(path, "/")
+    -- Strip trailing slashes from address for now
+    address = self.rtrim(address, "/")
+    -- Join our final URL, which *must* have a trailing / (it's a URL)
+    -- This is where we deviate from getJoinedPath ;).
+    local webdav_url = address .. "/" .. path
+    if webdav_url:sub(-1) ~= "/" then
         webdav_url = webdav_url .. "/"
     end
 
@@ -122,24 +102,19 @@ function WebDavApi:listFolder(address, user, pass, folder_path, folder_mode)
             --logger.dbg("WebDav catalog item=", item)
             -- <d:href> is the path and filename of the entry.
             local item_fullpath = item:match("<[^:]*:href[^>]*>(.*)</[^:]*:href>")
-            if string.sub( item_fullpath, -1 ) == "/" then
-                item_fullpath = string.sub( item_fullpath, 1, -2 )
-            end
-            local is_current_dir = self:isCurrentDirectory( util.urlDecode(item_fullpath), address, folder_path )
-
-            local item_name = util.urlDecode( FFIUtil.basename( item_fullpath ) )
-            item_name = util.htmlEntitiesToUtf8(item_name)
+            local item_name = FFIUtil.basename(util.htmlEntitiesToUtf8(util.urlDecode(item_fullpath)))
+            local is_current_dir = item_name == string.sub(folder_path, -#item_name)
 
             local is_not_collection = item:find("<[^:]*:resourcetype/>") or
                 item:find("<[^:]*:resourcetype></[^:]*:resourcetype>")
 
-            local item_path = (path == "" and has_trailing_slash) and item_name or path .. "/" .. item_name
+            local item_path = path .. "/" .. item_name
             if item:find("<[^:]*:collection[^<]*/>") then
                 item_name = item_name .. "/"
                 if not is_current_dir then
                     table.insert(webdav_list, {
                         text = item_name,
-                        url = util.urlDecode( item_path ),
+                        url = item_path,
                         type = "folder",
                     })
                 end
@@ -147,7 +122,7 @@ function WebDavApi:listFolder(address, user, pass, folder_path, folder_mode)
                 or G_reader_settings:isTrue("show_unsupported")) then
                 table.insert(webdav_file, {
                     text = item_name,
-                    url = util.urlDecode( item_path ),
+                    url = item_path,
                     type = "file",
                 })
             end
```

**Comment by mergen3107** [comment]
(don't know how to convert this to draft, closing for now)

**Comment by Frenzie** [comment]
I'm not sure if you can other than at the time of creation, maybe somewhere in the side menu.

**Comment by NiLuJe** [comment]
> I'm not sure if you can other than at the time of creation, maybe somewhere in the side menu.

Yep, it's at the top of the right side-penel, bundled with the reviews.

Rest assured, it takes me 35s to find it every time I look for it ;).

**Comment by benoit-pierre** [comment]
You can set it a creation? I always switch it on after open the PR…

**Comment by benoit-pierre** [comment]
Oh, right, the drop down on the 'Create pull request' button. TIL

**Comment by NiLuJe** [comment]
> Oh, right, the drop down on the 'Create pull request' button. TIL

Yeah, I also routinely forget where that one is hidden, although it is slightly more obvious ;).
